### DISCUSSION TITLE
duckdb.OutOfMemoryException: Out of Memory Error: Failed to allocate block of 16384 bytes

### DISCUSSION THREAD

**Discussion by KIC** [duckdb/duckdb#6904] [Q&A]
I have tried to load ~500k csv files using the `INSERT INTO xy SELECT * from 'file.csv'` syntax and a plain old for loop. But after more then 24h and only ~70k documents later it failed with duckdb.OutOfMemoryException: Out of Memory Error: Failed to allocate block of 16384 bytes

I am wondering how I not only can speed up the loading time but also manage to load all files. The compression rate looks really nice so far.

**Comment by Mytherin** [comment]
How large is each of the files? Are you loading into a persistent database, or into an in-memory database?

**Reply by KIC** [reply]
they have ~1000 records on average some are quite big actually. Yes of course a disk based database the raw csvs have almost 100GB (all together)

**Reply by Mytherin** [reply]
I would try batching together multiple files into a single transaction. 1000 records is a very low amount to load in a transaction and is likely the reason it is taking so long. You can read multiple files at once by using globs (e.g. `SELECT * FROM '*.csv'`), or you can manually construct a transaction by using `BEGIN` and `COMMIT`, e.g.:

```sql
BEGIN;
INSERT INTO xy SELECT * from 'f1.csv';
INSERT INTO xy SELECT * from 'f2.csv'
INSERT INTO xy SELECT * from 'f3.csv'
COMMIT;
```

I would aim for loading at least 1 million records per transaction.

**Reply by KIC** [reply]
Tanks, I will try batches of 1000.

I have a list of filenames which I need to filter i.e. by creation date, do I need to pass globs or can I pass a list of files?

**Reply by Mytherin** [reply]
You can pass a list of files to `read_csv_auto`, e.g.:

```sql
INSERT INTO xy SELECT * from read_csv_auto(['f1.csv', 'f2.csv'])
```

**Reply by KIC** [reply]
OK, currently I can commit 1000 files in ~75sec, it is for sure faster. Now lets hope it is not dying again

**Reply by KIC** [reply]
after 14 batches (14k dokuments) the time already doubles per batch now down to ~140secs. maybe it is just not the idea of duckdb to hold such an amount of data? roughly 500'000'000 records.

**Reply by KIC** [reply]
after 40k documents we are down to 3min per 1000 files. the db file is barely 7GB. I had sqlite files 10 times as big, so this shouldn't be a problem I guess

**Reply by KIC** [reply]
60k files loaded and now we are at 7min per 1000 files. it clearly degrades massively. also the CPU is much more under stress now. maybe the compression has some negative impact?

**Reply by Mytherin** [reply]
Hm, that is odd. Are you running individual `INSERT INTO` statements, or reading multiple files in a single `INSERT INTO`? How many rows are in the database? What are the data types of the table?

**Reply by KIC** [reply]
died again :-(

**Reply by KIC** [reply]
I have now concatenated all csv files into one giant gzip compressed csv. Now the `COPY` command seems to work fine (and amazingly fast) but I have somewhere a duplicate key violation. As it looks like from the [copy documentation](https://duckdb.org/docs/sql/statements/copy) there is no way to handle conflicts like `ON CONFLICT DO NOTHING`. Sadly the error does not even show which key is duplicated. 

However I have crated the same table without constraints and tried to figure out the duplicates using a query. But then again I get a memory issue:

```
D     create table foo as
>     select symbol, date, count(*)
>       from signal
>       group by symbol, date
>       having count(*) > 1;
 41% ▕████████████████████████▌                                   ▏ Error: Out of Memory Error: could not allocate block of 262144 bytes (26472747008/26472818278 used) 
```

**Reply by Mytherin** [reply]
Ah, that explains why you were running out of memory. [DuckDB does not currently support larger than memory indexes](https://duckdb.org/docs/sql/indexes).

I would just recommend not using a primary key index. DuckDB generally only uses indexes for point lookups (e.g. SELECT * FROM tbl WHERE id=X). Most analytical queries do not benefit from the index at all.

**Reply by KIC** [reply]
Since I could not load the csv file because of duplicate key issues my table has zero indexes at this point in time. So the select count(*) is actually running (and failing) on a non indexed table.

**Reply by Mytherin** [reply]
Yes, I meant running out of memory during the data load. `GROUP BY` will also currently not work if you have too many unique groups. In your case I assume that `symbol, date` is almost entirely unique (considering you mentioned encountering a duplicate only after loading many rows). That means `GROUP BY` on those two columns will essentially have the number of groups equivalent to the number of rows in the table, which will involve constructing a very large hash table that will likely fill up your RAM.

What I meant was just ignoring the duplicate and running queries as per usual (that likely don't involve billions of groups).

**Comment by l1t1** [comment]
how many distinct symbols in your table?
maybe you can partition it on symbol. 
then figure out duplicates on each partition.

**Reply by KIC** [reply]
I have about 400k unique symbols, which might be too many partitions. maybe I can partition by the date in ranges somehow? The only non float columns are `symbol VARCHAR` and `date TIMESTAMPTZ`.

**Reply by Mytherin** [reply]
You could partition on e.g. `YEAR(date)`, `MONTH(date)` or `DAY(date)` (or some combination of these).

**Comment by l1t1** [comment]
https://duckdb.org/docs/data/partitioning/partitioned_writes
